File: Python/Construct Binary Tree from Inorder and Postorder Traversal.py
"""
Construct Binary Tree from Postorder and Inorder Traversal.
"""

# Definition for a  binary tree node
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    # @param inorder, a list of integers
    # @param postorder, a list of integers
    # @return a tree node
    def buildTree(self, inorder, postorder):
        if len(inorder) == 0:
            return None
        root = TreeNode(postorder.pop())
        stn = list()
        stn.append(root)

        while True:
            if inorder[-1] == stn[-1].val:
                p = stn.pop()
                inorder.pop()
                if len(inorder) == 0: break
                if len(stn) != 0 and inorder[-1] == stn[-1].val: continue
                p.left = TreeNode(postorder.pop())
                stn.append(p.left)
            else:
                p = TreeNode(postorder.pop())
                stn[-1].right = p
                stn.append(p)

        return root

File: Java/Construct Binary Tree from Inorder and Postorder Traversal.java
/*
Given inorder and postorder traversal of a tree, construct the binary tree.

Note:
You may assume that duplicates do not exist in the tree.
*/

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public TreeNode buildTree(int[] inorder, int[] postorder) {
        if(inorder.length == 0) return null;
        
        TreeNode root = new TreeNode(postorder[postorder.length - 1]);
        Stack<TreeNode> stack = new Stack<>();
        stack.push(root);
        
        int idx1 = postorder.length - 2;
        int idx2 = inorder.length - 1;
        while(true){
            if(inorder[idx2] == stack.peek().val){
                TreeNode p = stack.pop();
                if(--idx2 < 0) break;
                if(!stack.isEmpty() && inorder[idx2] == stack.peek().val) continue;
                p.left = new TreeNode(postorder[idx1--]);
                stack.push(p.left);
            }
            else{
                TreeNode p = new TreeNode(postorder[idx1--]);
                stack.peek().right = p;
                stack.push(p);
            }
        }
        
        return root;
    }
}